pkg/ddl/partition.go

```diff
@@ -2872,6 +2936,33 @@ func (w *worker) onReorganizePartition(d *ddlCtx, t *meta.Meta, job *model.Job)
 			}
 		}
 
+		// All global indexes must be recreated, we cannot update them in-place, since we must have
+		// both old and new set of partition ids in the unique index at the same time!
+		for _, index := range tblInfo.Indices {
+			if !index.Unique {
+				// for now, only unique index can be global, non-unique indexes are 'local'
+				continue
+			}
+			inAllPartitionColumns, err := checkPartitionKeysConstraint(partInfo, index.Columns, tblInfo)
+			if err != nil {
+				return ver, errors.Trace(err)
+			}
+			if index.Global || !inAllPartitionColumns {
+				// Duplicate the unique indexes with new index ids.
+				// If previously was Global or will be Global:
+				// it must be recreated with new index ID
+				newIndex := *index
```

Comment:
It is better to use `index.Clone()`, otherwise `index.Columns` reference the same slice, which may be changed unexpectedly.

Reply:
Good catch, updated.